### DISCUSSION TITLE
Explanation of attributes

### DISCUSSION THREAD

**Discussion by tjafbe** [humbertogontijo/homeassistant-roborock#49] [General]
Since the integration is still fairly new, I thought we could maybe jointly start collecting some documentation info on what the specific attributes mean? Below is the list of attributes (in v0.0.17). I have filled in the ones I have personally been able to confirm with my own roborock. While I can certainly guess some of the others, I haven't been able to confirm them yet. When I do, I'll update the list. If you know what the attributes mean, let us know and I'll update the list accordingly.

| Attribute  | Explanation |
| ------------- | ------------- |
| Mop mode list |  List of possible mop modes (the routes)|
| Mop intensity list | List of possible mop intensity settings|
| Msg ver   | ? |
| Msg seq | ? |
| State      | number representing the vacuum states. But these are decoded in the 'State' and 'Vacuum status' attributes (see below) |
| Battery   | State of battery charge |
| Clean time | Time spent in cleaning on current job (in seconds) |
| Clean area | Cleaned area of the current job expressed in mm2. Divide by 1000000 to get it in m2 |
| Error code | ?|
| Map present | ?|
| In cleaning | not cleaning or spot cleaning -> in_cleaning: 0<br> full clean -> in_cleaning: 1<br>zoned clean -> in_cleaning: 2<br>room clean -> in_cleaning: 3|
| In returning | 1 when returning to dock, 0 otherwise |
| In fresh state | was 1 prior to leaving the dock and after the mop was cleaned and re-docked to empty the dust bin. It was 0 otherwise.|
| Lab status | ? |
| Water box status | 1 if the water box is attached otherwise 0 |
| Back type            | was 1 when returning to the dock (going_to_wash_the_mop), 0 when returning to the dock (going_home) and -1 otherwise. |
| Wash phase       | went from 11 to 13 to 12 while the mop was being washed (this was an automatic return to dock to clean the mop during a segment cleaning). It was 0 otherwise. |
| Wash ready        | ? |
| Fan power          | Vacuum suction power. Possible values are: 101 (silent), 102 (balanced), 103 (turbo), 104 (max), 108 (max+), 105 (off - mop only)|
| Dnd enabled | Status of "Do not disturb" mode. 0 is mode disabled, 1 means mode is enabled <br> <br>For dnd_enabled, which comes from the get_status return, does not change to 1 immediately if "Do Not Disturb Mode" is enabled in the app. There is another JSON response that comes from get_dnd_timer which has a key called enabled and that changes to 1 when it's enabled from the Roborock app almost immediately (on next update from the Roborock servers). It seems to take a while for dnd_enabled to change to 1 (maybe over an hour?) |
| Map status      | a number representing which floor map is loaded. This number does not seem to correspond to the number used for the command load_multi_map to actually change to different floor map. When I load map 0 with the load_multi_map command, this number changes to 3. When I load map 1, this number changes to 7, when I load map 2, this number changes to 11.  In any case, this attribute can be used to figure out which map is loaded.<br>255: when it is positioning. |
| Is locating | 1 when the vacuum is in "positioning" mode. When it loses its position, for example when you take it to a different floor, the vacuum will say "positioning". When it was able to find it's position, this flag will turn off.  |
| Lock status | ? It does NOT seem to be linked to child lock. |
| Water box mode | Mop scrub intensity. Possible values are: 200 (off), 201 (low), 202 (medium), 203 (high) |
| Water box carriage status | 1 if the mop is attached otherwise 0 |
| Mop forbidden enable | ? Something related to no mopping zones? |
| Camera status | seems to be values that are added with more options enabled. Here's what I observed. Remove viewing is disabled for all of these. I basically went down the options in the settings page and enabled one thing at a time, so each line has the previous option(s) enabled. <br><br> ReactiveAI Obstacle Avoidance - Off: 2448 <br> ReactiveAI Obstacle Avoidance - On: 2449 <br> AI Environment Recognition - On: 2449 (no change) <br> Less collision mode - On: 2449 (no change) <br> Obstacle photos - On: 3473 <br> Pet Details - Yes: 3475 <br> Fill Light - Automatic: 3459|
| Is exploring | ?|
| Home sec status | ?|
| Home sec enable password | ?|
| Adbumper status | ?|
| Water shortage status | _1 if there is no water in the water box otherwise 0_ <br> possibly incorrect. Comment below if you can confirm.|
| Dock type | 0: basic charging only dock <br>1: ?<br>2:?<br>3: the ultra dock also known as the "Empty Wash Fill dock" |
| Dust collection status | ?|
| Auto dust collection | ?|
| Avoid count | ?|
| Mop mode | Mop route setting. Possible values are: 300 (standard), 301 (deep), 303 (deep_plus) |
| Debug mode | ?|
| Collision avoid status | 1 is for the "Less collision mode" being enabled under the "ReactiveAI Obstacle Avoidance" settings page. |
| Switch map mode | ?|
| Dock error status | 0: no errors<br>38: when clean water tank is empty or removed (only updates when cleaning is started)<br>39: when dirty water tank is full or removed (only updates when cleaning is started)|
| Charge status | ?|
| Unsave map reason | ?|
| Unsave map flag | ?|
| State      | Current state of the Roborock. Possible values are: docked, cleaning, returning to dock |
| Vacuum status      | More detailed status, including things like returning to wash the mop etc. Notice that when docked, this will say charging, even when the battery is 100%. |
| Mop mode | Mop route setting. Possible values are: standard, deep, deep_plus, custom |

**Comment by shred86** [comment]
Observations from a Roborock S7 MaxV Ultra:

`collision_avoid_status: 1` is for the "Less collision mode" being enabled under the "ReactiveAI Obstacle Avoidance" settings page.

`camera_status` seems to be values that are added with more options enabled. Here's what I observed. Remove viewing is disabled for all of these. I basically went down the options in the settings page and enabled one thing at a time, so each line has the previous option(s) enabled.

- ReactiveAI Obstacle Avoidance - Off: 2448
- ReactiveAI Obstacle Avoidance - On: 2449
- AI Environment Recognition - On: 2449 (no change)
- Less collision mode - On: 2449 (no change)
- Obstacle photos - On: 3473
- Pet Details - Yes: 3475
- Fill Light - Automatic: 3459


Also, for `dnd_enabled`, which comes from the `get_status` return, does not change to `1` immediately if "Do Not Disturb Mode" is enabled in the app. There is another JSON response that comes from `get_dnd_timer` which has a key called `enabled` and that changes to `1` when it's enabled from the Roborock app almost immediately (on next update from the Roborock servers). It seems to take a while for `dnd_enabled` to change to 1 (maybe over an hour?).

**Reply by shred86** [reply]
`auto_dust_collection: 1` is when "Dock Settings" -> "Auto Emptying" is enabled.

I did not see any changes when the "Empty Mode" was changed in the app. Same with "Auto Drying".

I only recently got my S7 MaxV Ultra and my `avoid_count` is showing as 34, which I think makes sense with how many objects I've seen the camera recognize on the map but that's purely anecdotal.

**Reply by shred86** [reply]
More observations when I did a two room cleaning (vacuum & mop at the same time). Hard to pin point exactly what some of these do, especially with the 30 sec polling interval because the state could be changing multiple times within a 30 sec period. Anyways, just posting this up for reference. 
- `in_fresh_state:` was `1` prior to leaving the dock and after the mop was cleaned and re-docked to empty the dust bin. It was `0` otherwise.
- `back_type` was `1` when returning to the dock (going_to_wash_the_mop), `0` when returning to the dock (going_home) and `-1` otherwise.
- `wash_phase:` went from `11` to `13` to `12` while the mop was being washed (this was an automatic return to dock to clean the mop during a segment cleaning). It was `0` otherwise.
- `in_returning:` was `1` anytime the vacuum was going back to the dock; `0` otherwise.
- `unsave_map_reason:` was `6` while returning to the dock and `0` in all other cases (even after being docked).
- `avoid_count:` was `34` prior to this cleaning session but changed to `0` while the vacuum was cleaning and changed to `75` when returning to the dock. The map only showed it avoided a few objects during this cleaning session, so not sure what's going on.

**Comment by tjafbe** [comment]
Nice. Thanks for the help. I added this info to the table.

**Comment by jdejaegh** [comment]
Here are my observations on a S7 with v 0.0.11.

The `in_cleaning` is more subtle than what is currently in the description.  I tried multiple cleaning modes and the `in_cleaning` values changes according to the mode.
spot cleaning -> `in_cleaning: 0`
full clean -> `in_cleaning: 1`
zoned clean -> `in_cleaning: 2`
room clean -> `in_cleaning: 3`

The `mop_forbidden_enable` is 1 when the mop is installed on my vacuum, possibly to tell it to avoid the no mop zones.  It stays to 1 even if I remove all the no mop zones on the map or if I disable carpet detection.

Below are other observations I could not generalize enough but may be helpful to others.

I tried to guess what the `lock_status`.  I tried activating the child lock on the vacuum but this did not change the value.

My S7 with the basic charging dock always shows `dock_type: 0`

During my observations, `msg_ver: 2` never changed.  Given the name, I guess it's the protocol version used.  The `msg_seq` only increased, like a message counter or a sequence number (guessing from the name of the attribute).

I never saw another value than 0 for `water_shortage_status`, even when the water level is low or the water tank is removed.  I doubt that the S7 has such sensor.

**Reply by tjafbe** [reply]
Thanks. I added the relevant information. 

I think you are right about the water shortage. I have the ultra dock and when I emptied the clean water tank, the water_shortage_status stayed to 0 although the app gave me a warning that the clean water tank was empty. However I noticed that the dock_error_status changed to 38.

**Reply by jdejaegh** [reply]
Thanks! 
The `in_cleaning` is also set to 0 when not cleaning (e.g. idle, charging).

**Reply by shred86** [reply]
Seeing the same with the S7 MaxV. During segment cleaning (aka room cleaning), I was seeing `in_cleaning: 3` and when docked it was `0`. Also tried messing with the child lock but no change to `lock_status`. I'm showing `dock_type: 3` with the Empty Wash Fill dock. And `mop_forbidden_enable: 1` is showing with just 1 No-Go zone setup.

**Comment by falcon791** [comment]
`Dock type: 1` appears to be the Auto-Empty dock station (https://roborock.com.au/products/robodock-auto-empty-dock)

`Auto dust collection` seems to be the Auto-Empty settings for a compatible dock
`0` : Auto-Empty disabled
`1` : Auto Empty enabled, Empty Mode: doesn't appear to matter

I'm going to assume `Dust collection status` is `0` or `1` when emptying the vacuum but haven't been able to confirm this yet
